### PR TITLE
Document how to migrate an IML manager node to a different server

### PR DESCRIPTION
**Pull request by johnsonw**
Document how to migrate an IML manager node to a different server while still utilizing the existing MDS and OSS servers.

Signed-off-by: Will Johnson <william.c.johnson@intel.com>

### CODE REVIEW

File: docs/support/TOC.md [hunk 1]
```diff
@@ -20,6 +20,8 @@ This is the Support table of contents. Quickly find an issue that has been solve
   * If an OST can't be added there could be several reasons. Check here to see common reasons and their solutions.
 * [Re-add managed ZFS filesystem (3.1.x only)](re-add-managed-zfs-fs.md)
   * If you need to re-add a managed ZFS filesystem over an existing configuration. This procedure is meant for a fs that has the same targets and pool / dataset names, but different guids.
+* [Migrate an IML node to Existing ZFS monitored filesystem](migrate-iml-node-to-existing-zfs-monitored.md)
```
Comment (brianjmurrell) [inline docs/support/TOC.md]: Should say which kind of "IML node" you are referring to here.

The language isn't quite right in this description.  The "to Existing ZFS monitored filesystem" wants to indicate where you are doing the move to but you are not actually moving the node to the "ZFS monitored filesystem".

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
+<a name="top"></a>
+[Support Table of Contents](TOC.md)
+# Migrating an IML node to and existing ZFS Monitored Filesystem
```
Comment (brianjmurrell) [inline docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Ditto language.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+It may be necessary to migrate an IML manager node to a new server while still using an existing ZFS Monitored filesystem with the existing MDS and OSS servers. In this case, the IML manager node will be re-installed on a new server and will be setup to communicate with the existing MDS and OSS servers.
+
+## Performing The Migration
+
+### Backup the database on the existing server
+
+There are two backup methods:
+* Backup the entire database:
+```
```
Comment (brianjmurrell) [inline docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Set the type of code block here to `shell`.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+1. ssh into the manager node as root
+2. su -l postgres
+3. pg_dump -U chroma chroma > /tmp/db_backup_xxx.sql
+4. exit
```
Comment (brianjmurrell) [inline docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Let's remove the line numbers.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+```
+* Backup the database without metric data (Useful if backing up the entire database is too large):
+```
```
Comment (brianjmurrell) [inline docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Ditto.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+1. ssh into the manager node as root
+2. su -l postgres
+3. pg_dump -U chroma -F p -w -T 'chroma_core_series*' -T 'chroma_core_sample*' -T 'chroma_core_logmessage*' -f /tmp/db_backup_xxx.sql
+4. exit
```
Comment (brianjmurrell) [inline docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Ditto.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+[top](#top)
+
+### Setup server networking.
+
+The new server will need to communicate with the existing MDS and OSS nodes. Ensure the following:
+  * New IML Server Node:
+    * Verify that all MDS and OSS nodes are defined with the correct hostname and IP address in the /etc/hosts file. 
+  * Existing MDS and OSS nodes
+    * Verify that the IP address and hostname for the new IML server are reflected in the /etc/hosts file
+
```
Comment (brianjmurrell) [inline docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: What if the site isn't using `/etc/hosts` files, as any site with even a reasonable level of administration wouldn't be.

Maybe reword this to be more generic with "name resolution system (i.e. /etc/hosts, DNS, NIS, etc.).

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+[top](#top)
+
+
+### Setup ssh between the new IML server and the existing MDS and OSS nodes
+
+It is important that the IML manager node can communicate with the existing MDS and OSS nodes without having to enter a password. SSH keys should be setup to ensure that the manager node can ssh into any of the MDS/OSS nodes without having to specify a password. Likewise, all MDS and OSS nodes should be able to ssh into the new IML manager node. 
+
```
Comment (brianjmurrell) [inline docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Is password-less really a requirement?  I know we have options for the administrator to supply a password when adding hosts.  What if that's the security policy at the site?

Is ssh even necessary beyond this procedure?  Once IML has been set up on a cluster, we use the RPCs to tasks, not ssh.  ssh is only used for the initial node deploy normally.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+[top](#top)
+
+### Run the database import script.
+
+SCP the sql file from step 1 and the [import database script](scripts/import-customer-database.md) onto the new IML server. Once uploaded, ssh into the new IML server and execute the script, passing the sql filename as an argument:
+```
+scp import_database.py root@<new-iml-server-ip>:~
+scp chromadb_backup_xxx.sql root@<new-iml-server-ip>:~
+ssh root@<new-iml-server-ip>
```
Comment (brianjmurrell) [inline docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Let's make all of this more generic and not refer to ssh/scp directly since a site might have other tools they use to copy files around and log into remote hosts.

Wording like "copy foo to bar" and "on bar, execute...". Let them worry about the details of how to copy and access shells on their nodes.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+
+This will import the existing database file onto the new IML server and restart the chroma-manager service. 
+
+[top](#top)
+
+### Run the sha256 migration script.
+
+Download, extract, and run the appropriate version of the sha-256 migration script:
+* [sha-256-migration-IML-2.4.x](scripts/sha-256-migration/sha-256-browser-migration-2.4.x.v1.tar.gz)
+* [sha-256-migration-IML-3.x.x](scripts/sha-256-migration/sha-256-browser-migration-3.x.x.v1.tar.gz)
```
Comment (brianjmurrell) [inline docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Do we really want to be putting and storing scripts (never mind tarballs!) in the help documents/project?  Seems like whatever is in those tarballs should be in a github project so that their source is tracked and easily visible.  GH would also be the point of download in that case.

Should be more explicit about which scripts are for which version of IML.  What about 4.x?

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+```
+scp sha-256-browser-migration-3.x.x.v1.tar.gz root@<new-iml-server-ip>:~
+ssh root@<new-iml-server-ip>
```
Comment (brianjmurrell) [inline docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Make these more generic, not referring to ssh.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+https://<hostname|IP>:443/ui/
+```
+
+### Update the MDS and OSS nodes to point to the new IML server
+
+**If the hostname did not change then this step is not needed.** In the event that the hostname did change, the MDS and OSS nodes need to be configured such that they point to the new hostname. To accomplish this, the chroma settings on each of the server nodes will need to be updated:
+```
+# Example for MDS1
+cd /var/lib/chroma/settings
+for i in `ls`
```
Comment (brianjmurrell) [inline docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Use `$()` instead of backticks (``) for execution.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+```
+bGFzdF9kZXRlY3Rfc2Nhbl90YXJnZXRfZGV2aWNlcw==: last_detect_scan_target_devices
+c2VydmVy: server
+cHJvZmlsZQ==: profile
+YWdlbnQ=: agent
+```
+
+The file that maps to `server` needs to be updated to point to the new IML server hostname. For example, if the hostname changed from `adm.lfs.local` to `adm-centos72.lfs.local`,`c2VydmVy` would be updated as follows:
+```
+{"url": "https://adm-centos72.lfs.local:443/agent/"}
```
Comment (brianjmurrell) [inline docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: So, this line needs to be put into the file named `c2VydmVy`, replacing a similar line in that file?  Isn't that (base64 encoded) filename static?  IOW, won't it be the same name on every deployment?  If so, do we need to go through the process of base64 decoding any of the names instead of just telling them to update the file `c2VydmVy` with the new server JSON?

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md (hunk 10 above)
Comment (johnsonw) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: We do not have a migration script for 4.x as it uses sha-256 encryption instead of sh-1 encryption. Where should we store these files? It used to be stored on a portal somewhere but all of that has changed since we've moved to an open source model. We definitely need a place to store these files though, perhaps we can discuss in the IML eng. channel.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md (hunk 12 above)
Comment (johnsonw) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Sure. What's the advantage of `$()` over ````?

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md (hunk 13 above)
Comment (johnsonw) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: So I was wondering this myself. It sounds like it will be the same so i'll remove the for loop above.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md (hunk 7 above)
Comment (johnsonw) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: 👍

File: docs/support/TOC.md (hunk 1 above)
Comment (johnsonw) [reply docs/support/TOC.md]: How does, "Migrate an IML node to a new server while preserving an existing ZFS monitored filesystem" sound?

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md (hunk 10 above)
Comment (johnsonw) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: https://github.com/intel-hpdd/intel-manager-for-lustre/issues/514

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md (hunk 8 above)
Comment (johnsonw) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: I'll reword this accordingly.

Comment (johnsonw) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Actually, I can probably remove this section entirely.

File: docs/support/TOC.md (hunk 1 above)
Comment (brianjmurrell) [reply docs/support/TOC.md]: _Migrate an IML Manager configuration to a new server while preserving an existing ZFS monitored filesystem_

I do wonder what is ZFS specific about the procedure though.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md (hunk 12 above)
Comment (brianjmurrell) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: For one, it's more clear in documentation than a backtick which depending on font face, size, etc. can be unclear to distinguish from a single quote or an apostrophe,

But more importantly it's just a better version of `` that allows nesting.

Lots of useful discussion [here](https://stackoverflow.com/questions/9449778/what-is-the-benefit-of-using-instead-of-backticks-in-shell-scripts).

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md (hunk 13 above)
Comment (brianjmurrell) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: It seems that way to me.  It would be useful to get another opinion on this.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md [hunk 14]
```diff
@@ -49,13 +42,10 @@ It is important that the IML manager node can communicate with the existing MDS
...
 ### Run the database import script.
 
-SCP the sql file from step 1 and the [import database script](scripts/import-customer-database.md) onto the new IML server. Once uploaded, ssh into the new IML server and execute the script, passing the sql filename as an argument:
-```
-scp import_database.py root@<new-iml-server-ip>:~
-scp chromadb_backup_xxx.sql root@<new-iml-server-ip>:~
-ssh root@<new-iml-server-ip>
-[~]# chmod +x import_database.py
-[~]# ./import_database.py db_backup_xxx.sql
+Upload the sql file from step 1 and the [import database script](scripts/import-customer-database.md) onto the new IML server. Once uploaded, log into the new IML server and execute the script, passing the sql filename as an argument:
```
Comment (brianjmurrell) [inline docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Where is _scripts/import-customer-database.md_?  Is keeping it in the help system the right place?  I'm on the fence about this one.  If we decide to keep it in here, it probably shouldn't be `.md` though, right?

IML _manager_ server.

Comment (johnsonw) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: I guess this is something we need to discuss as a team. The support repo is intended to be a resource of information as well as scripts. The page can be found here: https://intel-hpdd.github.io/Support/docs/support/scripts/import-customer-database.html.

File: docs/support/TOC.md [hunk 15]
```diff
@@ -20,6 +20,8 @@ This is the Support table of contents. Quickly find an issue that has been solve
   * If an OST can't be added there could be several reasons. Check here to see common reasons and their solutions.
 * [Re-add managed ZFS filesystem (3.1.x only)](re-add-managed-zfs-fs.md)
   * If you need to re-add a managed ZFS filesystem over an existing configuration. This procedure is meant for a fs that has the same targets and pool / dataset names, but different guids.
+* [Migrate an IML node to a new server while preserving an existing ZFS monitored filesystem](migrate-iml-node-to-existing-zfs-monitored.md)
```
Comment (jgrund) [inline docs/support/TOC.md]: Does this need to be specific to ZFS?

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+su -l postgres
+pg_dump -U chroma chroma > /tmp/db_backup_xxx.sql
+exit
+```
+* Backup the database without metric data (Useful if backing up the entire database is too large):
+```shell
+# login to the manager node as root
+su -l postgres
+pg_dump -U chroma -F p -w -T 'chroma_core_series*' -T 'chroma_core_sample*' -T 'chroma_core_logmessage*' -f /tmp/db_backup_xxx.sql
+exit
```
Comment (jgrund) [inline docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Why would you want to backup the database without stats if you are migrating?

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md (hunk 10 above)
Comment (jgrund) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: I think having support-related code inline with the support docs is fine, as long as they are compartmentalized properly.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+Restarting Manager:
+Stopping daemons
+Starting daemons
+```
+
+[top](#top)
+
+### Refresh the browser and accept the certificate. Make sure the browser is pointed at the home page:
+```
+https://<hostname|IP>:443/ui/
```
Comment (jgrund) [inline docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: 443 is unnecessary.

File: docs/support/TOC.md (hunk 15 above)
Comment (johnsonw) [reply docs/support/TOC.md]: It will probably work for ldiskfs as well but I have not tested it yet.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md (hunk 16 above)
Comment (johnsonw) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: You would almost always want to backup the entire database, but I have run into issues in the past where the database file was so large we were not able to import it. This is simply showing the solution in case they run into this problem.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md (hunk 17 above)
Comment (johnsonw) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: i'll remove.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md (hunk 16 above)
Comment (brianjmurrell) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Well, if there is an issue importing, the solution isn't to just import only some of the data but rather address the problem with the import.  As long as there is enough disk space, there should not be any arbitrary limit on the size of the import.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+* [sha-256-migration-IML-3.x.x](scripts/sha-256-migration/sha-256-browser-migration-3.x.x.v1.tar.gz)
+
+```shell
+tar xvf sha-256-browser-migration-3.x.x.v1.tar.gz
+cd sha-256-browser-migration-3.x.x.v1
+./main
+```
+
+You should have output similar to the following:
+```
```
[the remark this answers is not in the document]
Comment (johnsonw) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Awesome! Thanks for sharing. I've downloaded it and using it now.

File: docs/support/TOC.md (hunk 1 above)
Comment (johnsonw) [reply docs/support/TOC.md]: I'll remove ZFS from the description.

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md (hunk 12 above)
Comment (johnsonw) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: Thanks for the clarification.

File: docs/support/TOC.md (hunk 15 above)
Comment (johnsonw) [reply docs/support/TOC.md]: I've removed all references to ZFS

File: docs/support/migrate-iml-node-to-existing-zfs-monitored.md (hunk 13 above)
Comment (jgrund) [reply docs/support/migrate-iml-node-to-existing-zfs-monitored.md]: It's static as long as the config is static in the agent; I don't see this changing anytime soon.

File: docs/support/migrate-iml-node-to-existing-monitored-fs.md [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+
+The new server will need to communicate with the existing MDS and OSS nodes. Ensure the following:
+  * New IML Server Node:
+    * Verify that all MDS and OSS nodes are defined with the correct hostname and IP address in the name resolution system (i.e. /etc/hosts, DNS, NIS, etc.).
+  * Existing MDS and OSS nodes
+    * Verify that the IP address and hostname for the new IML server are reflected in the name resolution system (i.e. /etc/hosts, DNS, NIS, etc.).
+
+[top](#migrate-an-iml-manager-configuration-to-a-new-server-while-preserving-an-existing-monitored-filesystem)
+
+### Download and Install IML.
```
Comment (jgrund) [inline docs/support/migrate-iml-node-to-existing-monitored-fs.md]: Might be worth providing a release page link here for the OSS releases.
